// koishi-plugin-new-repeater/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apply = exports.Config = exports.description = exports.name = void 0;
const koishi_1 = require("koishi");
exports.name = 'new-repeater';
exports.description = '复读机器人';
exports.Config = koishi_1.Schema.object({
    content: koishi_1.Schema.string().description('复读内容：').required(),
    debug: koishi_1.Schema.boolean().description('是否启动调试模式').default(false),
    possibility: koishi_1.Schema.number().description('复读概率').default(100),
});
function apply(ctx, config) {
    const logger = ctx.logger('new-repeater'); // 创建logger用于输出Koishi日志
    var ifDebug = false;
    ifDebug = config["debug"];
    function debug(content) {
        if (ifDebug)
            logger.info(content);
    }
    const content = config["content"];
    debug(`复读机启动完毕，复读内容：${content}`);
    ctx.on('message', (session) => {
        debug(`收到消息：${session.content}`);
        if (session.content === content) {
            // 随机复读
            let random = Math.floor(Math.random() * 100);
            if (random > config["possibility"]) {
                debug(`懒得复读：${session.content}`);
            }
            else {
                debug(`正在复读：${session.content}`);
                session.send(content);
            }
        }
    });
}
exports.apply = apply;
