// @minatojs/driver-sqlite/lib/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  SQLiteDriver: () => SQLiteDriver,
  default: () => src_default
});
module.exports = __toCommonJS(src_exports);
var import_cosmokit2 = require("cosmokit");
var import_minato2 = require("minato");
var import_sql_utils2 = require("@minatojs/sql-utils");
var import_node_path = require("node:path");
var import_promises = require("node:fs/promises");
var import_node_module = require("node:module");
var import_sql = __toESM(require("@minatojs/sql.js"), 1);

// src/locales/en-US.yml
var en_US_default = { path: "Database path." };

// src/locales/zh-CN.yml
var zh_CN_default = { path: "数据库路径。" };

// src/builder.ts
var import_sql_utils = require("@minatojs/sql-utils");
var import_cosmokit = require("cosmokit");
var import_minato = require("minato");
var SQLiteBuilder = class extends import_sql_utils.Builder {
  constructor(driver, tables) {
    super(driver, tables);
    this.driver = driver;
    this.queryOperators.$regexFor = (key, value) => typeof value === "string" ? `${this.escape(value)} regexp ${key}` : value.flags?.includes("i") ? `regexp2(${key}, ${this.escape(value.input)}, 'i')` : `${this.escape(value.input)} regexp ${key}`;
    this.evalOperators.$if = (args) => `iif(${args.map((arg) => this.parseEval(arg)).join(", ")})`;
    this.evalOperators.$regex = ([key, value, flags]) => flags?.includes("i") || value instanceof RegExp && value.flags?.includes("i") ? `regexp2(${this.parseEval(value)}, ${this.parseEval(key)}, ${this.escape(flags ?? value.flags)})` : `regexp(${this.parseEval(value)}, ${this.parseEval(key)})`;
    this.evalOperators.$concat = (args) => `(${args.map((arg) => this.parseEval(arg)).join("||")})`;
    this.evalOperators.$modulo = ([left, right]) => `modulo(${this.parseEval(left)}, ${this.parseEval(right)})`;
    this.evalOperators.$log = ([left, right]) => (0, import_cosmokit.isNullable)(right) ? `log(${this.parseEval(left)})` : `log(${this.parseEval(left)}) / log(${this.parseEval(right)})`;
    this.evalOperators.$length = (expr) => this.createAggr(expr, (value) => `count(${value})`, (value) => this.isEncoded() ? this.jsonLength(value) : this.asEncoded(`iif(${value}, LENGTH(${value}) - LENGTH(REPLACE(${value}, ${this.escape(",")}, ${this.escape("")})) + 1, 0)`, false));
    this.evalOperators.$number = (arg) => {
      const type = import_minato.Type.fromTerm(arg);
      const value = this.parseEval(arg);
      const res = import_minato.Field.date.includes(type.type) ? `cast(${value} / 1000 as integer)` : `cast(${this.parseEval(arg)} as double)`;
      return this.asEncoded(`ifnull(${res}, 0)`, false);
    };
    const binaryXor = /* @__PURE__ */ __name((left, right) => `((${left} & ~${right}) | (~${left} & ${right}))`, "binaryXor");
    this.evalOperators.$xor = (args) => {
      const type = import_minato.Type.fromTerm(this.state.expr, import_minato.Type.Boolean);
      if (import_minato.Field.boolean.includes(type.type))
        return args.map((arg) => this.parseEval(arg)).reduce((prev, curr) => `(${prev} != ${curr})`);
      else
        return args.map((arg) => this.parseEval(arg)).reduce((prev, curr) => binaryXor(prev, curr));
    };
    this.transformers["bigint"] = {
      encode: (value) => `cast(${value} as text)`,
      decode: (value) => `cast(${value} as integer)`,
      load: (value) => (0, import_cosmokit.isNullable)(value) ? value : BigInt(value),
      dump: (value) => (0, import_cosmokit.isNullable)(value) ? value : `${value}`
    };
    this.transformers["binary"] = {
      encode: (value) => `hex(${value})`,
      decode: (value) => `unhex(${value})`,
      load: (value) => (0, import_cosmokit.isNullable)(value) || typeof value === "object" ? value : import_cosmokit.Binary.fromHex(value),
      dump: (value) => (0, import_cosmokit.isNullable)(value) || typeof value === "string" ? value : import_cosmokit.Binary.toHex(value)
    };
  }
  static {
    __name(this, "SQLiteBuilder");
  }
  escapeMap = {
    "'": "''"
  };
  escapePrimitive(value, type) {
    if (value instanceof Date)
      value = +value;
    else if (value instanceof RegExp)
      value = value.source;
    else if (import_cosmokit.Binary.is(value))
      return `X'${import_cosmokit.Binary.toHex(value)}'`;
    else if (import_cosmokit.Binary.isSource(value))
      return `X'${import_cosmokit.Binary.toHex(import_cosmokit.Binary.fromSource(value))}'`;
    return super.escapePrimitive(value, type);
  }
  createElementQuery(key, value) {
    if (this.isJsonQuery(key)) {
      return this.jsonContains(key, this.escape(value, "json"));
    } else {
      return `(',' || ${key} || ',') LIKE ${this.escape("%," + value + ",%")}`;
    }
  }
  createRegExpQuery(key, value) {
    if (typeof value !== "string" && value.flags?.includes("i")) {
      return `regexp2(${this.escape(typeof value === "string" ? value : value.source)}, ${key}, ${this.escape(value.flags)})`;
    } else {
      return `regexp(${this.escape(typeof value === "string" ? value : value.source)}, ${key})`;
    }
  }
  jsonLength(value) {
    return this.asEncoded(`json_array_length(${value})`, false);
  }
  jsonContains(obj, value) {
    return this.asEncoded(`json_array_contains(${obj}, ${value})`, false);
  }
  encode(value, encoded, pure = false, type) {
    return encoded ? super.encode(value, encoded, pure, type) : encoded === this.isEncoded() && !pure ? value : this.asEncoded(this.transform(`(${value} ->> '$')`, type, "decode"), pure ? void 0 : false);
  }
  createAggr(expr, aggr, nonaggr) {
    if (!this.state.group && !nonaggr) {
      const value = this.parseEval(expr, false);
      return `(select ${aggr((0, import_sql_utils.escapeId)("value"))} from json_each(${value}) ${(0, import_minato.randomId)()})`;
    } else {
      return super.createAggr(expr, aggr, nonaggr);
    }
  }
  groupArray(value) {
    const res = this.isEncoded() ? `('[' || group_concat(${value}) || ']')` : `('[' || group_concat(json_quote(${value})) || ']')`;
    return this.asEncoded(`ifnull(${res}, json_array())`, true);
  }
  transformJsonField(obj, path) {
    return this.asEncoded(`(${obj} -> '$${path}')`, true);
  }
};

// src/index.ts
var import_node_url = require("node:url");
var import_meta = {};
function getTypeDef({ deftype: type }) {
  switch (type) {
    case "primary":
    case "boolean":
    case "integer":
    case "unsigned":
    case "bigint":
    case "date":
    case "time":
    case "timestamp":
      return `INTEGER`;
    case "float":
    case "double":
    case "decimal":
      return `REAL`;
    case "char":
    case "string":
    case "text":
    case "list":
    case "json":
      return `TEXT`;
    case "binary":
      return `BLOB`;
    default:
      throw new Error(`unsupported type: ${type}`);
  }
}
__name(getTypeDef, "getTypeDef");
var SQLiteDriver = class extends import_minato2.Driver {
  static name = "sqlite";
  path;
  db;
  sql = new SQLiteBuilder(this);
  beforeUnload;
  _transactionTask;
  /** synchronize table schema */
  async prepare(table, dropKeys) {
    const columns = this._all(`PRAGMA table_info(${(0, import_sql_utils2.escapeId)(table)})`);
    const model = this.model(table);
    const columnDefs = [];
    const indexDefs = [];
    const alter = [];
    const mapping = {};
    let shouldMigrate = false;
    for (const key in model.fields) {
      if (!import_minato2.Field.available(model.fields[key])) {
        if (dropKeys?.includes(key))
          shouldMigrate = true;
        continue;
      }
      const legacy = [key, ...model.fields[key].legacy || []];
      const column = columns.find(({ name }) => legacy.includes(name));
      const { initial, nullable = true } = model.fields[key];
      const typedef = getTypeDef(model.fields[key]);
      let def = `${(0, import_sql_utils2.escapeId)(key)} ${typedef}`;
      if (key === model.primary && model.autoInc) {
        def += " NOT NULL PRIMARY KEY AUTOINCREMENT";
      } else {
        def += (nullable ? " " : " NOT ") + "NULL";
        if (!(0, import_cosmokit2.isNullable)(initial)) {
          def += " DEFAULT " + this.sql.escape(this.sql.dump({ [key]: initial }, model)[key]);
        }
      }
      columnDefs.push(def);
      if (!column) {
        alter.push("ADD " + def);
      } else {
        mapping[column.name] = key;
        shouldMigrate ||= column.name !== key || column.type !== typedef;
      }
    }
    if (model.primary && !model.autoInc) {
      indexDefs.push(`PRIMARY KEY (${this._joinKeys((0, import_cosmokit2.makeArray)(model.primary))})`);
    }
    if (model.unique) {
      indexDefs.push(...model.unique.map((keys) => `UNIQUE (${this._joinKeys((0, import_cosmokit2.makeArray)(keys))})`));
    }
    if (model.foreign) {
      indexDefs.push(...Object.entries(model.foreign).map(([key, value]) => {
        const [table2, key2] = value;
        return `FOREIGN KEY (\`${key}\`) REFERENCES ${(0, import_sql_utils2.escapeId)(table2)} (\`${key2}\`)`;
      }));
    }
    if (!columns.length) {
      this.logger.info("auto creating table %c", table);
      this._run(`CREATE TABLE ${(0, import_sql_utils2.escapeId)(table)} (${[...columnDefs, ...indexDefs].join(", ")})`);
    } else if (shouldMigrate) {
      for (const { name, type, notnull, pk, dflt_value: value } of columns) {
        if (mapping[name] || dropKeys?.includes(name))
          continue;
        let def = `${(0, import_sql_utils2.escapeId)(name)} ${type}`;
        def += (notnull ? " NOT " : " ") + "NULL";
        if (pk)
          def += " PRIMARY KEY";
        if (value !== null)
          def += " DEFAULT " + this.sql.escape(value);
        columnDefs.push(def);
        mapping[name] = name;
      }
      const temp = table + "_temp";
      const fields = Object.keys(mapping).map(import_sql_utils2.escapeId).join(", ");
      this.logger.info("auto migrating table %c", table);
      this._run(`CREATE TABLE ${(0, import_sql_utils2.escapeId)(temp)} (${[...columnDefs, ...indexDefs].join(", ")})`);
      try {
        this._run(`INSERT INTO ${(0, import_sql_utils2.escapeId)(temp)} SELECT ${fields} FROM ${(0, import_sql_utils2.escapeId)(table)}`);
        this._run(`DROP TABLE ${(0, import_sql_utils2.escapeId)(table)}`);
      } catch (error) {
        this._run(`DROP TABLE ${(0, import_sql_utils2.escapeId)(temp)}`);
        throw error;
      }
      this._run(`ALTER TABLE ${(0, import_sql_utils2.escapeId)(temp)} RENAME TO ${(0, import_sql_utils2.escapeId)(table)}`);
    } else if (alter.length) {
      this.logger.info("auto updating table %c", table);
      for (const def of alter) {
        this._run(`ALTER TABLE ${(0, import_sql_utils2.escapeId)(table)} ${def}`);
      }
    }
    if (dropKeys)
      return;
    dropKeys = [];
    await this.migrate(table, {
      error: this.logger.warn,
      before: (keys) => keys.every((key) => columns.some(({ name }) => name === key)),
      after: (keys) => dropKeys.push(...keys),
      finalize: () => {
        if (!dropKeys.length)
          return;
        this.prepare(table, dropKeys);
      }
    });
  }
  async start() {
    this.path = this.config.path;
    if (this.path !== ":memory:") {
      this.path = (0, import_node_path.resolve)(this.ctx.baseDir, this.path);
    }
    const isBrowser = process.env.KOISHI_ENV === "browser";
    const sqlite = await (0, import_sql.default)({
      locateFile: (file) => process.env.KOISHI_BASE ? process.env.KOISHI_BASE + "/" + file : isBrowser ? "/modules/@koishijs/plugin-database-sqlite/" + file : (0, import_node_module.createRequire)(import_meta.url || (0, import_node_url.pathToFileURL)(__filename).href).resolve("@minatojs/sql.js/dist/" + file)
    });
    if (!isBrowser || this.path === ":memory:") {
      this.db = new sqlite.Database(this.path);
    } else {
      const buffer = await (0, import_promises.readFile)(this.path).catch(() => null);
      this.db = new sqlite.Database(this.path, buffer);
      if (isBrowser) {
        window.addEventListener("beforeunload", this.beforeUnload = () => {
          this._export();
        });
      }
    }
    this.db.create_function("regexp", (pattern, str) => +new RegExp(pattern).test(str));
    this.db.create_function("regexp2", (pattern, str, flags) => +new RegExp(pattern, flags).test(str));
    this.db.create_function("json_array_contains", (array, value) => +JSON.parse(array).includes(JSON.parse(value)));
    this.db.create_function("modulo", (left, right) => left % right);
    this.db.create_function("rand", () => Math.random());
    this.define({
      types: ["boolean"],
      dump: (value) => (0, import_cosmokit2.isNullable)(value) ? value : +value,
      load: (value) => (0, import_cosmokit2.isNullable)(value) ? value : !!value
    });
    this.define({
      types: ["json"],
      dump: (value) => JSON.stringify(value),
      load: (value) => typeof value === "string" ? JSON.parse(value) : value
    });
    this.define({
      types: ["list"],
      dump: (value) => Array.isArray(value) ? value.join(",") : value,
      load: (value) => value ? value.split(",") : []
    });
    this.define({
      types: ["date", "time", "timestamp"],
      dump: (value) => (0, import_cosmokit2.isNullable)(value) ? value : +new Date(value),
      load: (value) => (0, import_cosmokit2.isNullable)(value) ? value : new Date(Number(value))
    });
    this.define({
      types: ["binary"],
      dump: (value) => (0, import_cosmokit2.isNullable)(value) ? value : new Uint8Array(value),
      load: (value) => (0, import_cosmokit2.isNullable)(value) ? value : import_cosmokit2.Binary.fromSource(value)
    });
    this.define({
      types: import_minato2.Field.number,
      dump: (value) => value,
      load: (value) => (0, import_cosmokit2.isNullable)(value) ? value : Number(value)
    });
  }
  _joinKeys(keys) {
    return keys?.length ? keys.map((key) => `\`${key}\``).join(", ") : "*";
  }
  async stop() {
    await new Promise((resolve2) => setTimeout(resolve2, 0));
    this.db?.close();
    if (this.beforeUnload) {
      this.beforeUnload();
      window.removeEventListener("beforeunload", this.beforeUnload);
    }
  }
  _exec(sql, params, callback) {
    try {
      const stmt = this.db.prepare(sql);
      const result = callback(stmt);
      stmt.free();
      this.logger.debug("> %s", sql, params);
      return result;
    } catch (e) {
      this.logger.warn("> %s", sql, params);
      throw e;
    }
  }
  _all(sql, params = [], config) {
    return this._exec(sql, params, (stmt) => {
      stmt.bind(params);
      const result = [];
      while (stmt.step()) {
        result.push(stmt.getAsObject(null, config));
      }
      return result;
    });
  }
  _get(sql, params = [], config) {
    return this._exec(sql, params, (stmt) => stmt.getAsObject(params, config));
  }
  _export() {
    const data = this.db.export();
    return (0, import_promises.writeFile)(this.path, data);
  }
  _run(sql, params = [], callback) {
    this._exec(sql, params, (stmt) => stmt.run(params));
    const result = callback?.();
    return result;
  }
  async drop(table) {
    this._run(`DROP TABLE ${(0, import_sql_utils2.escapeId)(table)}`);
  }
  async dropAll() {
    const tables = Object.keys(this.database.tables);
    for (const table of tables) {
      this._run(`DROP TABLE ${(0, import_sql_utils2.escapeId)(table)}`);
    }
  }
  async stats() {
    const stats = { size: this.db.size(), tables: {} };
    const tableNames = this._all('SELECT name FROM sqlite_master WHERE type="table" ORDER BY name;');
    const dbstats = this._all('SELECT name, pgsize as size FROM "dbstat" WHERE aggregate=TRUE;');
    tableNames.forEach((tbl) => {
      stats.tables[tbl.name] = this._get(`SELECT COUNT(*) as count FROM ${(0, import_sql_utils2.escapeId)(tbl.name)};`);
      stats.tables[tbl.name].size = dbstats.find((o) => o.name === tbl.name).size;
    });
    return stats;
  }
  async remove(sel) {
    const { query, table, tables } = sel;
    const builder = new SQLiteBuilder(this, tables);
    const filter = builder.parseQuery(query);
    if (filter === "0")
      return {};
    const result = this._run(`DELETE FROM ${(0, import_sql_utils2.escapeId)(table)} WHERE ${filter}`, [], () => this._get(`SELECT changes() AS count`));
    return { matched: result.count, removed: result.count };
  }
  async get(sel) {
    const { model, tables } = sel;
    const builder = new SQLiteBuilder(this, tables);
    const sql = builder.get(sel);
    if (!sql)
      return [];
    const rows = this._all(sql, [], { useBigInt: true });
    return rows.map((row) => builder.load(row, model));
  }
  async eval(sel, expr) {
    const builder = new SQLiteBuilder(this, sel.tables);
    const inner = builder.get(sel.table, true, true);
    const output = builder.parseEval(expr, false);
    const { value } = this._get(`SELECT ${output} AS value FROM ${inner}`, [], { useBigInt: true });
    return builder.load(value, expr);
  }
  _update(sel, indexFields, updateFields, update, data) {
    const { ref, table, tables, model } = sel;
    const builder = new SQLiteBuilder(this, tables);
    (0, import_minato2.executeUpdate)(data, update, ref);
    const row = builder.dump(data, model);
    const assignment = updateFields.map((key) => `${(0, import_sql_utils2.escapeId)(key)} = ?`).join(",");
    const query = Object.fromEntries(indexFields.map((key) => [key, row[key]]));
    const filter = builder.parseQuery(query);
    this._run(`UPDATE ${(0, import_sql_utils2.escapeId)(table)} SET ${assignment} WHERE ${filter}`, updateFields.map((key) => row[key] ?? null));
  }
  async set(sel, update) {
    const { model, table, query } = sel;
    const { primary } = model, fields = model.avaiableFields();
    const updateFields = [...new Set(Object.keys(update).map((key) => {
      return Object.keys(fields).find((field) => field === key || key.startsWith(field + "."));
    }))];
    const primaryFields = (0, import_cosmokit2.makeArray)(primary);
    if (query.$expr || (0, import_minato2.hasSubquery)(sel.query) || Object.values(update).some((x) => (0, import_minato2.hasSubquery)(x))) {
      const sel2 = this.database.select(table, query);
      sel2.tables[sel.ref] = sel2.tables[sel2.ref];
      delete sel2.tables[sel2.ref];
      sel2.ref = sel.ref;
      const project = (0, import_cosmokit2.mapValues)(update, (value, key) => () => (0, import_minato2.isEvalExpr)(value) ? value : import_minato2.Eval.literal(value, model.getType(key)));
      const rawUpsert = await sel2.project({
        ...project,
        // do not touch sel2.row since it is not patched
        ...Object.fromEntries(primaryFields.map((x) => [x, () => (0, import_minato2.Eval)("", [sel.ref, x], sel2.model.getType(x))]))
      }).execute();
      const upsert = rawUpsert.map((row) => ({
        ...(0, import_cosmokit2.mapValues)(update, (_, key) => (0, import_minato2.getCell)(row, key)),
        ...Object.fromEntries(primaryFields.map((x) => [x, (0, import_minato2.getCell)(row, x)]))
      }));
      return this.database.upsert(table, upsert);
    } else {
      const data = await this.database.get(table, query);
      for (const row of data) {
        this._update(sel, primaryFields, updateFields, update, row);
      }
      return { matched: data.length };
    }
  }
  _create(table, data) {
    const model = this.model(table);
    data = this.sql.dump(data, model);
    const keys = Object.keys(data);
    const sql = `INSERT INTO ${(0, import_sql_utils2.escapeId)(table)} (${this._joinKeys(keys)}) VALUES (${Array(keys.length).fill("?").join(", ")})`;
    return this._run(sql, keys.map((key) => data[key] ?? null), () => this._get(`SELECT last_insert_rowid() AS id`));
  }
  async create(sel, data) {
    const { model, table } = sel;
    const { id } = this._create(table, data);
    const { autoInc, primary } = model;
    if (!autoInc || Array.isArray(primary))
      return data;
    return { ...data, [primary]: id };
  }
  async upsert(sel, data, keys) {
    if (!data.length)
      return {};
    const { model, table, ref } = sel;
    const fields = model.avaiableFields();
    const result = { inserted: 0, matched: 0, modified: 0 };
    const dataFields = [...new Set(Object.keys(Object.assign({}, ...data)).map((key) => {
      return Object.keys(fields).find((field) => field === key || key.startsWith(field + "."));
    }))];
    let updateFields = (0, import_cosmokit2.difference)(dataFields, keys);
    if (!updateFields.length)
      updateFields = [dataFields[0]];
    const step = Math.floor(960 / keys.length);
    for (let i = 0; i < data.length; i += step) {
      const chunk = data.slice(i, i + step);
      const results = await this.database.get(table, {
        $or: chunk.map((item) => Object.fromEntries(keys.map((key) => [key, item[key]])))
      });
      for (const item of chunk) {
        const row = results.find((row2) => {
          row2 = model.format(row2);
          return keys.every((key) => (0, import_cosmokit2.deepEqual)(row2[key], item[key], true));
        });
        if (row) {
          this._update(sel, keys, updateFields, item, row);
          result.matched++;
        } else {
          this._create(table, (0, import_minato2.executeUpdate)(model.create(), item, ref));
          result.inserted++;
        }
      }
    }
    return result;
  }
  async withTransaction(callback) {
    if (this._transactionTask)
      await this._transactionTask.catch(() => {
      });
    return this._transactionTask = new Promise((resolve2, reject) => {
      this._run("BEGIN TRANSACTION");
      callback().then(
        () => resolve2(this._run("COMMIT")),
        (e) => (this._run("ROLLBACK"), reject(e))
      );
    });
  }
};
((SQLiteDriver2) => {
  SQLiteDriver2.Config = import_minato2.z.object({
    path: import_minato2.z.string().role("path").required()
  }).i18n({
    "en-US": en_US_default,
    "zh-CN": zh_CN_default
  });
})(SQLiteDriver || (SQLiteDriver = {}));
var src_default = SQLiteDriver;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SQLiteDriver
});
